datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -120,6 +151,56 @@ use super::AggregateExec;
 /// hash table).
 ///
 /// [`group_values`]: Self::group_values
+///
+/// # Spilling
+///
+/// The sizes of group values and accumulators can become large. Before that causes out of memory,
+/// this hash aggregator outputs those data early for partial aggregation or spills to local disk
+/// using Arrow IPC format for final aggregation. For every input [`RecordBatch`], the memory
+/// manager checks whether the new input size meets the memory configuration. If not, outputting or
+/// spilling happens. For outputting, the final aggregation takes care of re-grouping. For spilling,
```

Comment:
How does it decide to output or spill?

Reply:
Looks you already got it, but just for record. A few lines above, I think this line addresses your question
```
this hash aggregator outputs those data early for partial aggregation or spills to local disk
using Arrow IPC format for final aggregation.
```